### PR TITLE
[typing/static] telemetry

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

Pure typing PR for telemetry code.

### How I Tested These Changes

BK

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12522** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12522" target="_blank">Graphite</a> 
    * **PR #12523** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12523" target="_blank">Graphite</a> 
      * **PR #12524** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12524" target="_blank">Graphite</a> 
        * **PR #12553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12553" target="_blank">Graphite</a> 
          * **PR #12714** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12714" target="_blank">Graphite</a>  👈
        * **PR #12713** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12713" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12714?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/telemetry.py [hunk 1]
```diff
@@ -76,8 +92,26 @@
     "BUILDKITE",  # https://buildkite.com/docs/pipelines/environment-variables
 }
 
+P = ParamSpec("P")
+T = TypeVar("T")
+T_Callable = TypeVar("T_Callable", bound=Callable[..., Any])
+
+TelemetryWrapperFn: TypeAlias = Callable[[T_Callable], T_Callable]
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/telemetry.py:99]: Doesn't seem like this is used anywhere?

Comment (smackesey) [reply python_modules/dagster/dagster/_core/telemetry.py]: good catch

File: python_modules/dagster/dagster/_core/telemetry.py [hunk 2]
```diff
@@ -77,8 +93,24 @@
     "BUILDKITE",  # https://buildkite.com/docs/pipelines/environment-variables
 }
 
+P = ParamSpec("P")
+T = TypeVar("T")
```
Comment (dpeng817) [inline python_modules/dagster/dagster/_core/telemetry.py:97]: would be nice to have some documentation here, or aliases for the chains of callables. Even as someone who has dealt a lot with the telemetry code, finding it very difficult to figure out what I'm looking at.

**Review by dpeng817 — CHANGES_REQUESTED** [review]
Would ideally like things to be a bit more readable around `@telemetry_wrapper` for example.

File: python_modules/dagster/dagster/_core/telemetry.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/telemetry.py:97]: Yeah this is a bit dense-- the signatures of higher order decorators always are. I think it's considerably clearere when we use the style of our user-facing higher-order decorators, separating the `target_fn` and `metadata` args.

**Review by dpeng817 — APPROVED** [review]
I think the target_fn bit makes things clearer - appreciated

**Pull request merged**